### ISSUE TITLE
vdirsyncer throws errors when syncing with one calendar

### ISSUE THREAD

**Issue by nielsk** [untitaker/vdirsyncer#225]
With one calendar I get the following errors:
These are the errors at the end when it is set to local wins:
```
warning: Race condition detected with server p01-caldav.icloud.com, consider using an alternative.
Copying (uploading) item 8374708D-8CF1-4A04-89CE-BA00005328659FEF to private_calendar_local
Unhandled exception occured for private_calendar.
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/vdirsyncer/storage/filesystem.py", line 141, in _upload_impl
    return fpath, get_etag_from_fileobject(f)
  File "/usr/lib/python3.4/contextlib.py", line 66, in __exit__
    next(self.gen)
  File "/usr/lib/python3.4/site-packages/atomicwrites/__init__.py", line 120, in _open
    self.commit(f)
  File "/usr/lib/python3.4/site-packages/atomicwrites/__init__.py", line 140, in commit
    move_atomic(f.name, self._path)
  File "/usr/lib/python3.4/site-packages/atomicwrites/__init__.py", line 72, in move_atomic
    return _move_atomic(src, dst)
  File "/usr/lib/python3.4/site-packages/atomicwrites/__init__.py", line 26, in _move_atomic
    os.link(src, dst)
FileExistsError: [Errno 17] File exists: '/home/nik/.khal/calendars/home/tmpygrlp2rd' -> '/home/nik/.khal/calendars/home/8374708D-8CF1-4A04-89CE-BA00005328659FEF.ics'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/vdirsyncer/cli/tasks.py", line 77, in sync_collection
    force_delete=force_delete
  File "/usr/lib/python3.4/site-packages/vdirsyncer/sync.py", line 175, in sync
    action(a_info, b_info, conflict_resolution)
  File "/usr/lib/python3.4/site-packages/vdirsyncer/sync.py", line 197, in inner
    dest_href, dest_etag = dest.storage.upload(item)
  File "/usr/lib/python3.4/site-packages/vdirsyncer/storage/base.py", line 17, in inner
    return f(self, *args, **kwargs)
  File "/usr/lib/python3.4/site-packages/vdirsyncer/storage/filesystem.py", line 118, in upload
    fpath, etag = self._upload_impl(item, href)
  File "/usr/lib/python3.4/site-packages/vdirsyncer/storage/filesystem.py", line 144, in _upload_impl
    raise exceptions.AlreadyExistingError(existing_href=href)
vdirsyncer.exceptions.AlreadyExistingError
```

And this is what I get when I set that remote wins
```
Doing conflict resolution for item 9B46D59C-D719-467E-800E-1E9445244877...
...private_calendar_remote wins.
Copying (updating) item 9B46D59C-D719-467E-800E-1E9445244877 to private_calendar_local
Copying (uploading) item 8374708D-8CF1-4A04-89CE-BA0000005328659FEF to private_calendar_local
Unhandled exception occured for private_calendar.
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/vdirsyncer/storage/filesystem.py", line 141, in _upload_impl
    return fpath, get_etag_from_fileobject(f)
  File "/usr/lib/python3.4/contextlib.py", line 66, in __exit__
    next(self.gen)
  File "/usr/lib/python3.4/site-packages/atomicwrites/__init__.py", line 120, in _open
    self.commit(f)
  File "/usr/lib/python3.4/site-packages/atomicwrites/__init__.py", line 140, in commit
    move_atomic(f.name, self._path)
  File "/usr/lib/python3.4/site-packages/atomicwrites/__init__.py", line 72, in move_atomic
    return _move_atomic(src, dst)
  File "/usr/lib/python3.4/site-packages/atomicwrites/__init__.py", line 26, in _move_atomic
    os.link(src, dst)
FileExistsError: [Errno 17] File exists: '/home/nik/.khal/calendars/home/tmp34i_i9w_' -> '/home/nik/.khal/calendars/home/8374708D-8CF1-4A04-89CE-BA0000005328659FEF.ics'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/vdirsyncer/cli/tasks.py", line 77, in sync_collection
    force_delete=force_delete
  File "/usr/lib/python3.4/site-packages/vdirsyncer/sync.py", line 175, in sync
    action(a_info, b_info, conflict_resolution)
  File "/usr/lib/python3.4/site-packages/vdirsyncer/sync.py", line 197, in inner
    dest_href, dest_etag = dest.storage.upload(item)
  File "/usr/lib/python3.4/site-packages/vdirsyncer/storage/base.py", line 17, in inner
    return f(self, *args, **kwargs)
  File "/usr/lib/python3.4/site-packages/vdirsyncer/storage/filesystem.py", line 118, in upload
    fpath, etag = self._upload_impl(item, href)
  File "/usr/lib/python3.4/site-packages/vdirsyncer/storage/filesystem.py", line 144, in _upload_impl
    raise exceptions.AlreadyExistingError(existing_href=href)
vdirsyncer.exceptions.AlreadyExistingError
```

When local wins no file-fixing (TZOFFSETFROM is wrong +5328 instead of +005328; a script I got from @untitaker fixes it for me) is needed, when remote wins file-fixing is needed (which is obvious). The problem exists with only one calendar, but there with lots of events.

This is how my vdirsyncer-config looks like:
```
[general]
status_path = ~/.vdirsyncer/status/
[pair private_calendar]
a = private_calendar_local
b = private_calendar_remote
conflict_resolution = b wins

[storage private_calendar_local]
type = filesystem
path = ~/.khal/calendars/home
fileext = .ics
[storage private_calendar_remote]
type = caldav
url = https://p01-caldav.icloud.com/this/is/a_path/
auth = guess
username = user@domain.tld
password = yeah_right
```

And this is the corresponding part in the khal.conf:
```
path = ~/.khal/khal.db
[calendars]
[[home]]
path = ~/.khal/calendars/home
color = light green
[locale]
local_timezone = Europe/Berlin
default_timezone = Europe/Berlin

timeformat = %H:%M
dateformat = %d.%m.
longdateformat = %d.%m.%Y
datetimeformat = %d.%m. %H:%M
longdatetimeformat = %d.%m.%Y %H:%M

firstweekday = 0

[default]
default_command = calendar
default_calendar = home
```

And this is the event that is shown in the errors above (there are far more but I hope one event is enough as a sample) after the script ran that fixes the wrong TZOFFSETFROM.

```
BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//CALENDARSERVER.ORG//NONSGML Version 1//EN
BEGIN:VTIMEZONE
TZID:Europe/Berlin
X-LIC-LOCATION:Europe/Berlin
BEGIN:STANDARD
DTSTART:18930401T000000
RDATE;VALUE=DATE-TIME:18930401T000000
TZNAME:CEST
TZOFFSETFROM:+005328
TZOFFSETTO:+0100
END:STANDARD
BEGIN:DAYLIGHT
DTSTART:19160430T230000
RDATE;VALUE=DATE-TIME:19160430T230000
RDATE;VALUE=DATE-TIME:19400401T020000
RDATE;VALUE=DATE-TIME:19430329T020000
RDATE;VALUE=DATE-TIME:19460414T020000
RDATE;VALUE=DATE-TIME:19470406T030000
RDATE;VALUE=DATE-TIME:19480418T020000
RDATE;VALUE=DATE-TIME:19490410T020000
RDATE;VALUE=DATE-TIME:19800406T020000
TZNAME:CEST
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
END:DAYLIGHT
BEGIN:STANDARD
DTSTART:19161001T010000
RDATE;VALUE=DATE-TIME:19161001T010000
RDATE;VALUE=DATE-TIME:19421102T030000
RDATE;VALUE=DATE-TIME:19431004T030000
RDATE;VALUE=DATE-TIME:19441002T030000
RDATE;VALUE=DATE-TIME:19451118T030000
RDATE;VALUE=DATE-TIME:19461007T030000
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
END:STANDARD
BEGIN:DAYLIGHT
DTSTART:19170416T020000
RRULE:FREQ=YEARLY;UNTIL=19180415T010000Z;BYDAY=3MO;BYMONTH=4
TZNAME:CEST
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
END:DAYLIGHT
BEGIN:STANDARD
DTSTART:19170917T030000
RRULE:FREQ=YEARLY;UNTIL=19180916T010000Z;BYDAY=3MO;BYMONTH=9
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
END:STANDARD
BEGIN:DAYLIGHT
DTSTART:19440403T020000
RRULE:FREQ=YEARLY;UNTIL=19450402T010000Z;BYDAY=1MO;BYMONTH=4
TZNAME:CEST
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
END:DAYLIGHT
BEGIN:DAYLIGHT
DTSTART:19450524T020000
RDATE;VALUE=DATE-TIME:19450524T020000
RDATE;VALUE=DATE-TIME:19470511T030000
TZNAME:CEMT
TZOFFSETFROM:+0200
TZOFFSETTO:+0300
END:DAYLIGHT
BEGIN:DAYLIGHT
DTSTART:19450924T030000
RDATE;VALUE=DATE-TIME:19450924T030000
RDATE;VALUE=DATE-TIME:19470629T030000
TZNAME:CEST
TZOFFSETFROM:+0300
TZOFFSETTO:+0200
END:DAYLIGHT
BEGIN:STANDARD
DTSTART:19460101T000000
RDATE;VALUE=DATE-TIME:19460101T000000
RDATE;VALUE=DATE-TIME:19800101T000000
TZNAME:CEST
TZOFFSETFROM:+0100
TZOFFSETTO:+0100
END:STANDARD
BEGIN:STANDARD
DTSTART:19471005T030000
RRULE:FREQ=YEARLY;UNTIL=19491002T010000Z;BYDAY=1SU;BYMONTH=10
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
END:STANDARD
BEGIN:STANDARD
DTSTART:19800928T030000
RRULE:FREQ=YEARLY;UNTIL=19950924T010000Z;BYDAY=-1SU;BYMONTH=9
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
END:STANDARD
BEGIN:DAYLIGHT
DTSTART:19810329T020000
RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3
TZNAME:CEST
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
END:DAYLIGHT
BEGIN:STANDARD
DTSTART:19961027T030000
RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
END:STANDARD
END:VTIMEZONE
BEGIN:VEVENT
UID:8374708D-8CF1-4A04-89CE-BA000000000000005328659FEF
DTEND;TZID=Europe/Berlin:20081009T153000
SUMMARY:Fahrschule
DTSTART;TZID=Europe/Berlin:20081009T143000
DTSTAMP:20110908T161944Z
SEQUENCE:0
TRANSP:OPAQUE
END:VEVENT
END:VCALENDAR
```

As recommended in https://github.com/geier/khal/issues/140#issuecomment-118603502 I did a 
`vdirsyncer repair private_calendar_remote/home`

After that the problems persist. I also tried again with local/remote wins.
The repair worked on the items and didn't throw any error

**Comment by nielsk** [comment]
The vdirsyncer repair-command told me that it processes 1320 items. After a sync with remote-sync which needs fixing I count 1229 ics-files that get re-downloaded apparently

**Comment by untitaker** [comment]
Could you

* delete all local calendars
* delete the status folder
* `vdirsyncer sync`

**Comment by untitaker** [comment]
Also I've just fixed a bug in master where an execution of the repair command could've tripped up vdirsyncer massively.

**Comment by nielsk** [comment]
vdirsyncer sync worked
Then I ran the sync-script which fixes TZOFFSETFROM (for a lot of items on all calendars)
Then I ran vdirsyncer sync again in the hope that all the experimenting doesn't break my calendars. It synced everything back up with loads of Race conditions but you mentioned earlier that this warning happens too often apparently anyway.
2nd sync: still works
Added an event locally, added an event remotely. Both synced, fixes got synced and nothing broke apparently. Thanks a lot :)
